council/llm/llm_middleware.py

```diff
@@ -186,3 +189,102 @@ def __call__(self, llm: LLMBase, execute: ExecuteLLMRequest, request: LLMRequest
+
+
+class CacheEntry:
+    """Represents a cached LLM response."""
+
+    def __init__(self, response: LLMResponse, timestamp: float, ttl: float) -> None:
+        self.response = self._rebuild_response(response)
+        self.timestamp = timestamp
+        self.ttl = ttl
+
+    @staticmethod
+    def _rebuild_response(response: LLMResponse) -> LLMResponse:
+        """Modify the response so it has zero duration and consumption with different units"""
+        if response.result is not None:
+            cached_consumptions: List[Consumption] = [
+                Consumption(value=consumption.value, unit=f"cached_{consumption.unit}", kind=consumption.kind)
+                for consumption in response.result.consumptions
+            ]
+            cached_result = LLMResult(response.result.choices, cached_consumptions, response.result.raw_response)
+        else:
+            cached_result = None
+        cached_response = LLMResponse(response._request, cached_result, 0)
+
+        return cached_response
+
+    @property
+    def is_expired(self) -> bool:
+        return time.time() - self.timestamp >= self.ttl
+
+    def renew(self) -> None:
+        """Update the timestamp to extend the lifetime."""
+        self.timestamp = time.time()
+
+
+class LLMCachingMiddleware:
+    """Middleware that caches LLM responses to avoid duplicate calls."""
+
+    def __init__(self, ttl: float = 300.0, cache_limit_size: int = 10) -> None:
+        """
+        Initialize the caching middleware.
+
+        Args:
+            ttl: Time-to-live in seconds for cache entries (default: 5 mins)
+            cache_limit_size: Cache limit size in cache units (default: 10)
+        """
+        self._cache: OrderedDict[str, CacheEntry] = OrderedDict()
+        self._ttl = ttl
+        self.cache_limit_size = cache_limit_size
+
+    def __call__(self, llm: LLMBase, execute: ExecuteLLMRequest, request: LLMRequest) -> LLMResponse:
+        self._remove_expired()
+        key = self.get_request_hash(request)
+
+        if key in self._cache:
+            entry = self._cache[key]
+
+            if not entry.is_expired:
+                entry.renew()  # renew the entry lifetime on hit
+                self._cache.move_to_end(key)  # move to most recent
+                return entry.response
+
+            del self._cache[key]  # remove expired entry
+
+        response = execute(request)
+
+        # cache the new response
+        self._cache[key] = CacheEntry(response=response, timestamp=time.time(), ttl=self._ttl)
+        self._cache.move_to_end(key)
+        self._enforce_cache_limit()
+
+        return response
+
+    @staticmethod
+    def get_request_hash(request: LLMRequest) -> str:
+        """Convert the request to a hash with hashlib.sha256."""
+        serialized = json.dumps(
+            {
+                # TODO: request.context is not serialized
+                "messages": [str(m) for m in request.messages],  # TODO: serialize messages properly
```

Comment:
@Winston-503 do we want:
- to lower case
- suppress spaces
- any other optimization?

Reply:
I think lowercasing and removing whitespace characters is enough, but we should do it for `content` and also each `data` element